### PR TITLE
feat: type safety time

### PR DESCRIPTION
**Pull request by HoganEdwardChu**
Hi @rueian  for the issue of type safety i tried fix but not confident for the change can you look into this? Thanks!

### CODE REVIEW

**Comment by rueian** [comment]
Hi @HoganEdwardChu, thank you for the quick PR. However, we can’t break the current API. This have to be done by adding new functions instead of modifying the old ones. cc @SoulPancake

**Comment by HoganEdwardChu** [comment]
> Hi @HoganEdwardChu, thank you for the quick PR. However, we can’t break the current API. This have to be done by adding new functions instead of modifying the old ones. cc @SoulPancake

got it you mean add kind like ExSeconds() right?

**Comment by rueian** [comment]
`PxMillisecondsWithType` might be too long. I much prefer something like `Px`. However, I don't know if it is possible.

Another thing is that I don't think modifying hack/cmds/commands.json is a good idea. We will migrate to the new format of commands.json maintained by Redis org one day and such special changes will be lost and hard to notice.

**Comment by HoganEdwardChu** [comment]
> `PxMillisecondsWithType` might be too long. I much prefer something like `Px`. However, I don't know if it is possible.
> 
> Another thing is that I don't think modifying hack/cmds/commands.json is a good idea. We will migrate to the new format of commands.json maintained by Redis org one day and such special changes will be lost and hard to notice.

Oh... migrating to new format of commands.json... Maybe in this issue, we can modify commands.json and later i think. we can handle it for another way

**Comment by HoganEdwardChu** [comment]
> > Maybe in this issue, we can modify commands.json and make functions with type safety
> 
> Okay. But instead of introducing a whole new enum block in commands.json, you should add the four type-safe variants to the original expiration block. In that way, I believe it will much easier to notice these special modifications in commands.json one day when we migrate to the new format.

@rueian thanks i apply it

File: hack/cmds/gen.go [hunk 1]
```diff
@@ -817,6 +835,10 @@ func printBuilder(w io.Writer, parent, next goStruct) {
 						follows = append(follows, toGoName(p.Name)+"...")
 					case "...string": // TODO hack for FT.CREATE VECTOR
 						follows = append(follows, toGoName(p.Name)+"...")
+					case "time.Duration":
+						appends = append(appends, fmt.Sprintf("strconv.FormatInt(int64(%s/time.Millisecond), 10)", toGoName(p.Name)))
+					case "time.Time":
+						appends = append(appends, fmt.Sprintf("%s.Format(time.RFC3339)", toGoName(p.Name)))
```
[the remark this answers is not in the document]
Comment (HoganEdwardChu) [reply hack/cmds/gen.go:841]: oh i will fix it

File: hack/cmds/gen.go [hunk 2]
```diff
@@ -149,6 +149,16 @@ func (n *node) GoStructs() (out []goStruct) {
...
+					// FIXME: this should be handle differently later
+				case "sec-typed", "ms-typed":
+					// Using time.Duration for EX/PX
+					s.BuildDef.Command = append(s.BuildDef.Command, cmds[:1]...)
+					s.BuildDef.Parameters = []parameter{{Name: lcFirst(name(cmds[1])), Type: "time.Duration"}}
+					// FIXME: this should be handle differently later
+				case "timestamp-typed", "ms-timestamp-typed":
+					// Using time.Time for EXAT/PXAT
+					s.BuildDef.Command = append(s.BuildDef.Command, cmds[:1]...)
+					s.BuildDef.Parameters = []parameter{{Name: lcFirst(name(cmds[1])), Type: "time.Time"}}
```
[the remark this answers is not in the document]
Comment (HoganEdwardChu) [reply hack/cmds/gen.go]: oh got it

File: hack/cmds/gen.go [hunk 3]
```diff
@@ -836,9 +840,19 @@ func printBuilder(w io.Writer, parent, next goStruct) {
...
+						case next.BuildDef.MethodName == "Px":
+							appends = append(appends, fmt.Sprintf("strconv.FormatInt(int64(%s/time.Millisecond), 10)", toGoName(p.Name)))
+						}
 					case "time.Time":
-						appends = append(appends, fmt.Sprintf("%s.Format(time.RFC3339)", toGoName(p.Name)))
+						switch {
+						case next.BuildDef.MethodName == "Exat":
+							appends = append(appends, fmt.Sprintf("strconv.FormatInt(%s.Unix(), 10)", toGoName(p.Name))) // For seconds
+						case next.BuildDef.MethodName == "Pxat":
+							appends = append(appends, fmt.Sprintf("strconv.FormatInt(%s.UnixNano()/1e6, 10)", toGoName(p.Name))) // For milliseconds
```
Comment (rueian) [inline hack/cmds/gen.go:854]: Just use the `.UnixMilli()`

**Comment by HoganEdwardChu** [comment]
Seems almost done @rueian ... now we can have type-safety functions for ex, pxat ...

**Comment by rueian** [comment]
Many thanks @HoganEdwardChu !

**Pull request merged**